### PR TITLE
WT-17296 Merge cross checkpoint caching feature branch

### PR DESCRIPTION
**Pull request by zunyiliu**
This PR is to merge the cross checkpoint caching feature branch to develop. 
Despite the tickets PR, there're few changes to the branch to fix some bugs uncovered by evergreen patch:
1. [Remove the redundant assert on cache destroy which will fail on error path](https://github.com/wiredtiger/wiredtiger/pull/13810/changes/99adccaa94f6e563b22531ae32f8cc9d77c7f258), on error path we assert that's a disagg conn when destroy the shared disk cache, where it may have not initialised yet.
2. [Don't cache non-disag btree pages](https://github.com/wiredtiger/wiredtiger/pull/13810/changes/3cabbe7d3e30a1aeb22dcb341006af6d871a8cc1) In disagg we still have local file pages, like local metadata file, these pages should not have shared disk images.
3. [Fix btree bytes underflow bug](https://github.com/wiredtiger/wiredtiger/pull/13810/changes/ee51558f432e1685eeaeada6fa6172d7be7c1b72) We used to account disk bytes to btree bytes when the shared disk image is loaded for the first time, this is wrong as a shared disk image lifecycle is only controlled by shared disk cache layer, we may have loaded a disk image for a page on btree1, and there're few btree pages referencing the same disk image, it's not guaranteed the last page referencing that disk image will be released for the same btree. I fixed this by removing the btree bytes accounting for shared disk image.
4. [Disable shared disk cache for now as we need the full complete feature](https://github.com/wiredtiger/wiredtiger/pull/13810/changes/fbbc62410c8b254b3c5fcdaed4c1019ebe4dea77) though I have passed all wt tests when it's enabled, it fails 2 mongo-master/sys-perf tests. That may because of some rare edge case, but most likely it is due to the incomplete of the feature code, like we haven't supported step-down and step-up, and we are very conservative on bucket size/spin lock now, I'm not enabling it until the related tickets are merged in, otherwise we may see break out failures and many sys-perf regressions.

### CODE REVIEW

File: src/btree/bt_page.c [hunk 1]
```diff
@@ -1276,11 +1276,23 @@ __wti_page_inmem(WT_SESSION_IMPL *session, WT_REF *ref, const void *image, uint3
...
+     */
+    if (shared_dsk_item != NULL)
+        LF_SET(WT_PAGE_DISK_SHARED);
+
     /* Allocate and initialize a new WT_PAGE. */
     WT_RET(__wt_page_alloc(session, dsk->type, alloc_entries, true, &page, flags));
     __wt_tsan_suppress_store_wt_page_header_ptr(&page->dsk, dsk);
     F_SET_ATOMIC_16(page, flags);
 
+    /* Update image stats early so it's balanced with __wt_page_out decr if an error fires below. */
```
Comment (quchenhao) [inline src/btree/bt_page.c]: better to not mention function names in the comment. Better here it may easier to make the comments more understandable. So I'm ok to leave it here but up to you.

File: src/btree/bt_read.c [hunk 2]
```diff
@@ -385,27 +403,57 @@ __page_read(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags)
...
-          __wti_page_inmem(session, ref, new_image_copy.data, page_flags, &page, &instantiate_upd));
-    else {
-        WT_ERR(__wti_page_inmem(session, ref, tmp[0].data, page_flags, &page, &instantiate_upd));
+
+    disk_image_buf = build_full_disk_image_from_deltas ? &new_image_copy : &tmp[0];
+
+    if (shared_dsk_cache->enabled && F_ISSET(btree, WT_BTREE_DISAGGREGATED)) {
+        bool shared_dsk_inserted = false;
+
+        /* A cache hit takes the skip_disk_read path, so we can't already have an item here. */
```
Comment (quchenhao) [inline src/btree/bt_read.c]: better to not mention variable names in the comment

File: src/cache/cache.c [hunk 3]
```diff
@@ -64,12 +64,37 @@ __wt_cache_config(WT_SESSION_IMPL *session, const char *cfg[], bool reconfig)
...
 
     /* Use a common routine for run-time configuration options. */
     WT_RET(__wt_cache_config(session, cfg, false));
 
+    /*
+     * Initialize the shared disk hash table only on disaggregated nodes. Size the hash table at
+     * 0.2% of cache size divided by ~100B per entry (cache_size / 500 / 100), with a minimum of 512
+     * buckets.
+     *
+     * We cannot use __wt_conn_is_disagg here because __wti_disagg_conn_config hasn't run yet and
```
Comment (quchenhao) [inline src/cache/cache.c]: can you avoid mentioning function names in the comment?

File: src/cache/cache.c [hunk 4]
```diff
@@ -64,12 +64,37 @@ __wt_cache_config(WT_SESSION_IMPL *session, const char *cfg[], bool reconfig)
...
+     * page_log_meta is still NULL. Read the config directly instead.
+     *
+     * FIXME-WT-14721: Once the init ordering in conn_open.c is fixed so that
+     * __wti_disagg_conn_config runs before __wt_cache_create, replace this config lookup with
```
Comment (quchenhao) [inline src/cache/cache.c]: same here

**Review by aye-sha256 — COMMENTED** [review]
Discussed on Slack with @zunyiliu that the decoupling of the disk image from page/btree might affect the eviction walk threshold being met so a tree might get skipped when it otherwise wouldn't. Not within project scope and potentially might not even be a problem in performance but will make a ticket to keep in mind

**Comment by zunyiliu** [comment]
@quchenhao Thanks! I've fixed the comments in [this commit](https://github.com/wiredtiger/wiredtiger/pull/13810/changes/6426583318b71f2296b25a6c2cdd557bda53f67e).

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by aye-sha256 — APPROVED** [review]
